File: bindings/java/src/io/quad4/rns/Rsg.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

package io.quad4.rns;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rns.ffi.RnsLibrary;

/** RSG / RSM helpers. */
public final class Rsg {
    private Rsg() {}

    public static byte[] create(Identity identity, byte[] message, boolean embed) {
        byte[] payload = message == null ? new byte[0] : message;
        LongByReference needed = new LongByReference();
        int probe =
                RnsLibrary.INSTANCE.rns_rsg_create(
                        identity.handle(), payload, payload.length, embed ? 1 : 0, null, 0, needed);
        if (probe != RnsLibrary.RNS_OK && probe != RnsLibrary.RNS_ERR_TRUNCATED) {
            Rns.check(probe);
        }
        if (needed.getValue() == 0) {
            throw new RnsException(RnsException.INTERNAL);
        }
        byte[] out = new byte[(int) needed.getValue()];
        LongByReference written = new LongByReference();
        Rns.check(
                RnsLibrary.INSTANCE.rns_rsg_create(
                        identity.handle(),
                        payload,
                        payload.length,
                        embed ? 1 : 0,
                        out,
                        out.length,
                        written));
        byte[] result = new byte[(int) written.getValue()];
        System.arraycopy(out, 0, result, 0, result.length);
        return result;
    }

    public static void validate(byte[] rsg, byte[] message, byte[] requiredSignerHash) {
        byte[] r = rsg == null ? new byte[0] : rsg;
        byte[] m = message == null ? new byte[0] : message;
        byte[] s = requiredSignerHash == null ? new byte[0] : requiredSignerHash;
        Rns.check(RnsLibrary.INSTANCE.rns_rsg_validate(r, r.length, m, m.length, s, s.length));
    }

    public static byte[] rsmVerify(byte[] rsm, byte[] requiredSignerHash) {
        byte[] r = rsm == null ? new byte[0] : rsm;
        byte[] s = requiredSignerHash == null ? new byte[0] : requiredSignerHash;
        LongByReference needed = new LongByReference();
        int probe = RnsLibrary.INSTANCE.rns_rsm_verify(r, r.length, s, s.length, null, 0, needed);
        if (probe != RnsLibrary.RNS_OK && probe != RnsLibrary.RNS_ERR_TRUNCATED) {
            Rns.check(probe);
        }
        if (needed.getValue() == 0) {
            return new byte[0];
        }
        byte[] out = new byte[(int) needed.getValue()];
        LongByReference written = new LongByReference();
        Rns.check(RnsLibrary.INSTANCE.rns_rsm_verify(r, r.length, s, s.length, out, out.length, written));
        byte[] result = new byte[(int) written.getValue()];
        System.arraycopy(out, 0, result, 0, result.length);
        return result;
    }
}

File: bindings/lua/rns/rsg.lua
-- SPDX-License-Identifier: Apache-2.0
-- Copyright (c) 2024-2026 Quad4.io

local ffi_mod = require("rns.ffi")
local errors = require("rns.errors")
local ffi = ffi_mod.ffi
local lib = ffi_mod.lib

local function rsg_create(identity, message, embed)
	local msg, n = ffi_mod.to_uint8(message or "")
	local needed = ffi.new("size_t[1]", 0)
	local code = lib.rns_rsg_create(identity:handle(), msg, n, embed and 1 or 0, nil, 0, needed)
	if code ~= errors.Error.OK and code ~= errors.Error.TRUNCATED then
		errors.map_code(code)
	end
	if needed[0] == 0 then
		errors.raise(errors.Error.INTERNAL)
	end
	local out = ffi.new("uint8_t[?]", needed[0])
	local written = ffi.new("size_t[1]", 0)
	errors.map_code(lib.rns_rsg_create(identity:handle(), msg, n, embed and 1 or 0, out, needed[0], written))
	return ffi.string(out, written[0])
end

local function rsg_validate(rsg, message, required_signer_hash)
	local rbuf, rn = ffi_mod.to_uint8(rsg)
	local mbuf, mn = ffi_mod.to_uint8(message or "")
	local sbuf, sn = ffi_mod.to_uint8(required_signer_hash or "")
	errors.map_code(lib.rns_rsg_validate(rbuf, rn, mbuf, mn, sbuf, sn))
end

local function rsm_verify(rsm, required_signer_hash)
	local rbuf, rn = ffi_mod.to_uint8(rsm)
	local sbuf, sn = ffi_mod.to_uint8(required_signer_hash or "")
	local needed = ffi.new("size_t[1]", 0)
	local code = lib.rns_rsm_verify(rbuf, rn, sbuf, sn, nil, 0, needed)
	if code ~= errors.Error.OK and code ~= errors.Error.TRUNCATED then
		errors.map_code(code)
	end
	if needed[0] == 0 then
		return ""
	end
	local out = ffi.new("uint8_t[?]", needed[0])
	local written = ffi.new("size_t[1]", 0)
	errors.map_code(lib.rns_rsm_verify(rbuf, rn, sbuf, sn, out, needed[0], written))
	return ffi.string(out, written[0])
end

return {
	rsg_create = rsg_create,
	rsg_validate = rsg_validate,
	rsm_verify = rsm_verify,
}
